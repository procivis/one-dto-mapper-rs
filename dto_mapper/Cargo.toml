[package]
edition = "2021"
name    = "dto_mapper"
publish = false
version = "0.1.0"

[dependencies]
dto_mapper_derive = { version = "0.1", path = "../dto_mapper_derive" }
fmap              = { version = "0.8.3" }
