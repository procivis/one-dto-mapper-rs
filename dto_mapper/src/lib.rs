pub use dto_mapper_derive::*;
use fmap::Functor;

pub fn convert_inner<'a, T, A>(outer: T) -> T::Mapped
where
    T: Functor<'a, A>,
    T::Inner: Into<A>,
{
    outer.fmap(Into::into)
}

pub fn convert_inner_of_inner<'a, T, K, A: 'a>(outer: T) -> T::Mapped
where
    T: Functor<'a, K>,
    T::Inner: Functor<'a, A, Mapped = K>,
    <T::Inner as Functor<'a, A>>::Inner: Into<A>,
{
    outer.fmap(convert_inner)
}

pub fn iterable_try_into<T, C, R>(input: C) -> Result<R, <C::Item as TryInto<T>>::Error>
where
    C: IntoIterator,
    C::Item: TryInto<T>,
    R: FromIterator<T>,
{
    input.into_iter().map(|item| item.try_into()).collect()
}
