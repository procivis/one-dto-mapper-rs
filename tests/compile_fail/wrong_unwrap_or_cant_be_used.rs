use dto_mapper::From;

struct OptionalDto {
    age: Option<u16>,
}

#[derive(From)]
#[convert(from = NumericDto)]
struct StringDto {
    #[convert(into_unwrap_or = "16")]
    age: u16,
}

fn main() {}
