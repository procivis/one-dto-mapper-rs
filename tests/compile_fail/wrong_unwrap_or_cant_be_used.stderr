error: `into_unwrap_or` can be used only when `into` is specified
 --> tests/compile_fail/wrong_unwrap_or_cant_be_used.rs:7:10
  |
7 | #[derive(From)]
  |          ^^^^
  |
  = note: this error originates in the derive macro `From` (in Nightly builds, run with -Z macro-backtrace for more info)
